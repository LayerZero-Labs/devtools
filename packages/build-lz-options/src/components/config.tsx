import React from "react";
import { Box, Text } from "ink";
import {
  OptionType1Summary,
  OptionTypeInput,
  OptionType2Summary,
} from "@/types";

interface Props {
  value: OptionTypeInput;
}

interface OptionType1Props {
  props: OptionType1Summary;
}

interface OptionType2Props {
  props: OptionType2Summary;
}

export const ConfigSummary: React.FC<Props> = ({ value }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Creating LayerZero options <Text bold>{value.type.label}</Text>
      </Text>
    </Box>
  );
};

export const Option1Summary: React.FC<OptionType1Props> = ({ props }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Gas Limit: <Text bold>{props.gasLimit}</Text>
      </Text>
    </Box>
  );
};

export const Option2Summary: React.FC<OptionType2Props> = ({ props }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Gas Limit: <Text bold>{props.gasLimit}</Text>
      </Text>
      <Text>
        Native Drop Amount: <Text bold>{props.nativeDropAmount}</Text>
      </Text>
      <Text>
        Native Drop Address: <Text bold>{props.nativeDropAddress}</Text>
      </Text>
    </Box>
  );
};
