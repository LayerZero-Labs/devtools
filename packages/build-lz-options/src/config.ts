import type { OptionType } from '@/types'
import { ExecutorOptionType, WorkerId } from '@layerzerolabs/lz-v2-utilities'

/**
 * Supported Option Types.
 */
export const OPTION_TYPES: OptionType[] = [
    {
        // TODO: use OptionType.TYPE_1 once exported from lz-v2-utility
        id: '1',
        label: '1: gas for remote execution',
    },
    {
        // TODO: use OptionType.TYPE_2 once exported from lz-v2-utility
        id: '2',
        label: '2: gas for remote execution and native drop',
    },
    {
        // TODO: use OptionType.TYPE_3 once exported from lz-v2-utility
        id: '3',
        label: '3: options builder (EndpointV2 only)',
    },
]

/**
 * Supported Executor Option Types.
 */
export const EXECUTOR_OPTION_TYPE = [
    {
        id: ExecutorOptionType.LZ_RECEIVE,
        label: '1: lzReceive',
    },
    {
        id: ExecutorOptionType.NATIVE_DROP,
        label: '2: nativeDrop',
    },
    {
        id: ExecutorOptionType.COMPOSE,
        label: '3: compose',
    },
    {
        id: ExecutorOptionType.ORDERED,
        label: '4: ordered',
    },
]

/**
 * Supported Worker Types.
 */
export const WORKER_TYPE = [
    {
        id: WorkerId.EXECUTOR,
        label: '1 Executor',
    },
    {
        id: WorkerId.VERIFIER,
        label: '2 Verifier',
    },
]
