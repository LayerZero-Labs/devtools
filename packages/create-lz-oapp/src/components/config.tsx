import React from "react";
import { Box, Text } from "ink";
import { Config } from "@/types.js";
import { resolve } from "path";

interface Props {
  value: Config;
}

export const ConfigSummary: React.FC<Props> = ({ value }) => {
  return (
    <Box flexDirection="column" marginTop={1} marginBottom={1}>
      <Text>
        Will create a project in{" "}
        <Text bold>{value.destination || "current"}</Text> directory (
        <Text bold>{resolve(value.destination)}</Text>)
      </Text>
      <Text>
        Will use the <Text bold>{value.example.label}</Text> example
      </Text>
      <Text>
        Will use <Text bold>{value.packageManager.label}</Text> to install
        dependencies
      </Text>
    </Box>
  );
};
