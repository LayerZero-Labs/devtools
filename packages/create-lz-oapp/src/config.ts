import type { Example, PackageManager } from '@/types.js'

/**
 * To enable example development in a custom repository
 * we open the repository URL field to be taken from the environment
 */
const repository = process.env['LAYERZERO_EXAMPLES_REPOSITORY_URL'] || 'git@github.com:LayerZero-Labs/lz-utils'

/**
 * To enable example development in a custom branch,
 * we open up the ref field to be taken from the environment
 *
 * `LAYERZERO_EXAMPLES_REPOSITORY_REF` can then be set to something like `#develop` or `#my-custom-branch`
 * to take the examples from a tag, a branch or a commit hash
 */
const ref = process.env['LAYERZERO_EXAMPLES_REPOSITORY_REF'] || ''

export const EXAMPLES: Example[] = [
    {
        id: 'oft',
        label: 'OFT',
        repository,
        directory: 'examples/oft',
        ref,
    },
    {
        id: 'oapp',
        label: 'OApp',
        repository,
        directory: 'examples/oapp',
        ref,
    },
]

export const PACKAGE_MANAGERS: PackageManager[] = [
    {
        command: 'npm',
        label: 'npm',
    },
    {
        command: 'yarn',
        label: 'yarn',
    },
    {
        command: 'pnpm',
        label: 'pnpm',
    },
    {
        command: 'bun',
        label: 'bun',
    },
]
