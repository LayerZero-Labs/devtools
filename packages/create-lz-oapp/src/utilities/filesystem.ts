import { lstatSync } from 'fs'

export const isDirectory = (path: string) => {
    try {
        return lstatSync(path).isDirectory()
    } catch {
        return false
    }
}
