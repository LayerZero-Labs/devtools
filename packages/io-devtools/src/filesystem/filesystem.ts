import { createModuleLogger } from '@/stdio/logger'
import { accessSync, constants, lstatSync } from 'fs'

export const isDirectory = (path: string): boolean => {
    try {
        return lstatSync(path).isDirectory()
    } catch {
        return false
    }
}

export const isFile = (path: string): boolean => {
    try {
        return lstatSync(path).isFile()
    } catch {
        return false
    }
}

export const isReadable = (path: string): boolean => {
    try {
        return accessSync(path, constants.R_OK), true
    } catch {
        return false
    }
}

/**
 * CJS/ESM interoperable utility for importing JS/TS/JSON files
 *
 * This utility is not safe in a sense that it will not check for file existence
 * before trying to import it, consumers need to do that
 *
 * @param {string} path
 * @returns {unknown} imported module
 */
export const importDefault = async (path: string): Promise<unknown> => {
    const logger = createModuleLogger('filesystem')

    // Let's use a dynamic import first
    //
    // This will not work in new node versions without experimental ESM module support
    try {
        logger.debug(`Importing default from '${path}' using dynamic import`)
        const result = await import(path)

        if (result != null && 'default' in result) {
            return logger.debug(`Found default in '${path}' using dynamic import`), result.default
        }

        // If not let's just return the whole export
        return logger.debug(`Did not find default in '${path}' using dynamic import, returning the whole thing`), result
    } catch (error) {
        logger.debug(`Failed to import from '${path}' using dynamic import: ${error}`)
    }

    // If dynamic import fails, let's use good ol' require
    try {
        logger.debug(`Importing default from '${path}' using require`)
        const result = await require(path)

        if (result != null && result.__esModule) {
            return logger.debug(`Found default in '${path}' using require`), result.default
        }

        // If not let's just return the whole export
        return logger.debug(`Did not find default in '${path}' using require, returning the whole thing`), result
    } catch (error) {
        logger.debug(`Failed to import from '${path}' using require: ${error}`)

        // We'll just rethrow the second error
        throw error
    }
}
