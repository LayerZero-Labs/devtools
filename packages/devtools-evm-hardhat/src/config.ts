import 'hardhat-deploy/dist/src/type-extensions'

import { endpointIdToNetwork } from '@layerzerolabs/lz-definitions'
import { HardhatUserConfig } from 'hardhat/types'
import { join, dirname } from 'path'
import { createModuleLogger } from '@layerzerolabs/io-devtools'

const resolvePackageDirectory = (packageName: string): string => {
    // The tricky bit here is the fact that if we resolve packages by their package name,
    // we might be pointed to a file in some dist directory - node will just pick up the `main`
    // entry in package.json and point us there
    //
    // So in order to get a stable path we choose package.json, pretty solid choice
    const packageJsonName = join(packageName, 'package.json')
    // We now resolve the path to package.json
    const packageJsonPath = require.resolve(packageJsonName)
    // And return its directory
    return dirname(packageJsonPath)
}

/**
 * Helper utility that adds external deployment paths for all LayzerZero enabled networks.
 * This will make LayerZero contracts available in your deploy scripts and tasks.
 *
 * ```
 * // hardhat.config.ts
 * import { EndpointId } from "@layerzerolabs/lz-definitions"
 *
 * const config: HardhatUserConfig = {
 *   networks: {
 *     arbitrum: {
 *       eid: EndpointId.ARBITRUM_MAINNET
 *     },
 *     fuji: {
 *       eid: EndpointId.AVALANCHE_TESTNET
 *     }
 *   }
 * }
 *
 * export default withLayerZeroDeployments("@layerzerolabs/lz-evm-sdk-v1")
 * ```
 *
 * @param packageNames `string[]` List of @layerzerolabs package names that contain deployments directory
 *
 * @returns `<THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig` Hardhat config decorator
 */
export const withLayerZeroDeployments = (...packageNames: string[]) => {
    const resolvedDeploymentsDirectories = packageNames
        // The first thing we do is we resolve the paths to LayerZero packages
        .map(resolvePackageDirectory)
        // Then navigate to the deployments folder
        .map((resolvedPackagePath) => join(resolvedPackagePath, 'deployments'))

    // We return a function that will enrich hardhat config with the external deployments configuration
    //
    // This is a pretty standard way of enriching configuration files that leads to quite nice consumer code
    return <THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig => ({
        ...config,
        external: {
            ...config.external,
            // Now for the meat of the operation, we'll enrich the external.deployments object
            deployments: Object.fromEntries(
                Object.entries(config.networks ?? {}).flatMap(([networkName, networkConfig]) => {
                    const eid = networkConfig?.eid
                    const networkLogger = createModuleLogger(networkName)

                    // Let's first check whether eid is defined on the network config
                    if (eid == null) {
                        networkLogger.debug(
                            'Endpoint ID not specified in hardhat config, skipping external deployment configuration'
                        )

                        return []
                    }

                    try {
                        // This operation is unsafe and can throw - let's make sure we don't explode with some unreadable error
                        const layerZeroNetworkName = endpointIdToNetwork(eid)
                        const layerZeroNetworkDeploymentsDirectories = resolvedDeploymentsDirectories.map(
                            (deploymentsDirectory) => join(deploymentsDirectory, layerZeroNetworkName)
                        )

                        return [
                            [
                                // The external deployments object is keyed by local network names
                                // which do not necessarily match the LayerZero ones
                                networkName,
                                // And its values are arrays of filesystem paths referring to individual network deployment directories
                                Array.from(
                                    // Since we want the paths to be unique, we'll put everything we have into a Set, then convert back to array
                                    new Set([
                                        // These are the external deployments already configured
                                        ...(config.external?.deployments?.[networkName] ?? []),
                                        // And these are the new ones
                                        ...layerZeroNetworkDeploymentsDirectories,
                                    ])
                                ),
                            ],
                        ]
                    } catch (error) {
                        networkLogger.error(
                            `Invalid endpoint ID specified in hardhat config (${eid}), skipping external deployment configuration`
                        )

                        return []
                    }
                })
            ),
        },
    })
}

/**
 * Helper utility that appends external artifacts directories
 * to existing hadrhat config
 *
 * ```
 * // hardhat.config.ts
 * import { EndpointId } from "@layerzerolabs/lz-definitions"
 *
 * const config: HardhatUserConfig = {
 *   networks: {
 *     arbitrum: {
 *       eid: EndpointId.ARBITRUM_MAINNET
 *     },
 *     fuji: {
 *       eid: EndpointId.AVALANCHE_TESTNET
 *     }
 *   }
 * }
 *
 * export default withLayerZeroArtifacts("@layerzerolabs/lz-evm-sdk-v1")
 * ```
 *
 * @param packageNames `string[]`
 *
 * @returns `<THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig` Hardhat config decorator
 */
export const withLayerZeroArtifacts = (...packageNames: string[]) => {
    const resolvedArtifactsDirectories = packageNames
        // The first thing we do is we resolve the paths to LayerZero packages
        .map(resolvePackageDirectory)
        // Then navigate to the artifacts folder
        .map((resolvedPackagePath) => join(resolvedPackagePath, 'artifacts'))

    // We return a function that will enrich hardhat config with the external artifacts configuration
    //
    // This is a pretty standard way of enriching configuration files that leads to quite nice consumer code
    return <THardhatUserConfig extends HardhatUserConfig>(config: THardhatUserConfig): THardhatUserConfig => {
        // We'll first grab all the external artifacts already defined
        const existingArtifacts = new Set(config.external?.contracts?.flatMap(({ artifacts }) => artifacts) ?? [])

        // And only append stuff if we have something new to say
        const newArtifacts = new Set(
            resolvedArtifactsDirectories.filter((artifact) => !existingArtifacts.has(artifact))
        )
        if (newArtifacts.size === 0) {
            return config
        }

        return {
            ...config,
            external: {
                ...config.external,
                contracts: [
                    ...(config.external?.contracts ?? []),
                    {
                        artifacts: Array.from(newArtifacts),
                    },
                ],
            },
        }
    }
}
