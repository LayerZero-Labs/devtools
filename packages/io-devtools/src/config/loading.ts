import { importDefault, isFile, isReadable } from '@/filesystem/filesystem'
import { createModuleLogger } from '@/stdio/logger'
import { printZodErrors } from '@/stdio/printer'
import { resolve } from 'path'
import { z } from 'zod'

export const createConfigLoader =
    <TConfig>(schema: z.ZodSchema<TConfig>, logger = createModuleLogger('config loader')) =>
    async (path: string): Promise<TConfig> => {
        const absolutePath = resolve(path)
        logger.verbose(`Resolved config file location for '${path}': '${absolutePath}'`)

        // First we check that the config file is indeed there and we can read it
        logger.verbose(`Checking config file '${absolutePath}' for existence & readability`)
        const isConfigReadable = isFile(absolutePath) && isReadable(absolutePath)
        if (!isConfigReadable) {
            throw new Error(
                `Unable to read config file '${path}'. Check that the file exists and is readable to your terminal user`
            )
        }

        // Keep talking to the user
        logger.verbose(`Config file '${absolutePath}' exists & is readable`)

        // Now let's see if we can load the config file
        let rawConfig: unknown
        try {
            logger.verbose(`Loading config file '${absolutePath}'`)

            rawConfig = await importDefault(absolutePath)
        } catch (error) {
            throw new Error(`Unable to read config file '${path}': ${error}`)
        }

        logger.verbose(`Loaded config file '${absolutePath}'`)

        // Now let's check whether the config file contains a function
        //
        // If so, we'll execute this function and will expect a config as a result
        let rawConfigMaterialized: unknown
        if (typeof rawConfig === 'function') {
            logger.verbose(`Executing configuration function from config file '${absolutePath}'`)

            try {
                rawConfigMaterialized = await rawConfig()
            } catch (error) {
                throw new Error(`Got an exception while executing config funtion from file '${path}': ${error}`)
            }
        } else {
            logger.verbose(`Using exported value from config file '${absolutePath}'`)
            rawConfigMaterialized = rawConfig
        }

        // It's time to make sure that the config is not malformed
        //
        // At this stage we are only interested in the shape of the data,
        // we are not checking whether the information makes sense (e.g.
        // whether there are no missing nodes etc)
        logger.verbose(`Validating the structure of config file '${absolutePath}'`)

        const configParseResult = schema.safeParse(rawConfigMaterialized)
        if (configParseResult.success === false) {
            const userFriendlyErrors = printZodErrors(configParseResult.error)

            throw new Error(
                `Config from file '${path}' is malformed. Please fix the following errors:\n\n${userFriendlyErrors}`
            )
        }

        return configParseResult.data
    }
