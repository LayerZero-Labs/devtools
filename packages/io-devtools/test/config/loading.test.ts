import { createConfigLoader } from '@/config'
import { importDefault, isFile, isReadable } from '@/filesystem/filesystem'
import { z } from 'zod'

jest.mock('@/filesystem/filesystem', () => ({
    isFile: jest.fn().mockReturnValue(false),
    isReadable: jest.fn().mockReturnValue(false),
    importDefault: jest.fn().mockRejectedValue('Not mocked'),
}))

const isFileMock = isFile as jest.Mock
const importDefaultMock = importDefault as jest.Mock
const isReadableMock = isReadable as jest.Mock

describe('config/loading', () => {
    beforeEach(() => {
        isFileMock.mockReset()
        importDefaultMock.mockReset()
        isReadableMock.mockReset()
    })

    describe('createConfigLoader', () => {
        it('should reject if the path is not a file', async () => {
            isFileMock.mockReturnValue(false)

            const configLoader = createConfigLoader(z.unknown())

            await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()
        })

        it('should reject if the path is not readable', async () => {
            isFileMock.mockReturnValue(true)
            isReadableMock.mockReturnValue(false)

            const configLoader = createConfigLoader(z.unknown())

            await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()
        })

        it('should reject if the file cannot be imported', async () => {
            isFileMock.mockReturnValue(true)
            isReadableMock.mockReturnValue(true)
            importDefaultMock.mockRejectedValue('No way')

            const configLoader = createConfigLoader(z.unknown())

            await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()
        })

        describe('when config file exports a config', () => {
            it('should reject if the file contents do not match the schema', async () => {
                isFileMock.mockReturnValue(true)
                isReadableMock.mockReturnValue(true)
                importDefaultMock.mockResolvedValue({ bad: 'config' })

                const schema = z.object({ good: z.string() })
                const configLoader = createConfigLoader(schema)

                await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()
            })

            it('should resolve if the file contents match the schema', async () => {
                isFileMock.mockReturnValue(true)
                isReadableMock.mockReturnValue(true)
                importDefaultMock.mockResolvedValue({ good: 'config' })

                const schema = z.object({ good: z.string() })
                const configLoader = createConfigLoader(schema)

                await expect(configLoader('./myconfig.ts')).resolves.toMatchSnapshot()
            })
        })

        describe('when config file exports a function', () => {
            describe('when it is a synchronous function', () => {
                it('should reject if the function throws an error', async () => {
                    const configFunctionMock = jest.fn(() => {
                        throw new Error('Oh not again')
                    })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    const schema = z.object({ good: z.string() })
                    const configLoader = createConfigLoader(schema)

                    await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should reject if the function returns an invalid config', async () => {
                    const configFunctionMock = jest.fn().mockReturnValue({ bad: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    const schema = z.object({ good: z.string() })
                    const configLoader = createConfigLoader(schema)

                    await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should resolve if the function returns a valid config', async () => {
                    const configFunctionMock = jest.fn().mockReturnValue({ good: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    const schema = z.object({ good: z.string() })
                    const configLoader = createConfigLoader(schema)

                    await expect(configLoader('./myconfig.ts')).resolves.toEqual({ good: 'config' })

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })
            })

            describe('when it is an asynchronous function', () => {
                it('should reject if the function rejects', async () => {
                    const configFunctionMock = jest.fn().mockRejectedValue(new Error('Y u do dis'))

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    const schema = z.object({ good: z.string() })
                    const configLoader = createConfigLoader(schema)

                    await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should reject if the function resolves with an invalid config', async () => {
                    const configFunctionMock = jest.fn().mockResolvedValue({ bad: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    const schema = z.object({ good: z.string() })
                    const configLoader = createConfigLoader(schema)

                    await expect(configLoader('./myconfig.ts')).rejects.toMatchSnapshot()

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })

                it('should resolve if the function returns a valid config', async () => {
                    const configFunctionMock = jest.fn().mockResolvedValue({ good: 'config' })

                    isFileMock.mockReturnValue(true)
                    isReadableMock.mockReturnValue(true)
                    importDefaultMock.mockResolvedValue(configFunctionMock)

                    const schema = z.object({ good: z.string() })
                    const configLoader = createConfigLoader(schema)

                    await expect(configLoader('./myconfig.ts')).resolves.toEqual({ good: 'config' })

                    expect(configFunctionMock).toHaveBeenCalledTimes(1)
                    expect(configFunctionMock).toHaveBeenCalledWith()
                })
            })
        })
    })
})
