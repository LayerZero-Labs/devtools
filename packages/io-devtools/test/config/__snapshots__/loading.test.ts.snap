// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config/loading createConfigLoader should reject if the file cannot be imported 1`] = `[Error: Unable to read config file './myconfig.ts': No way]`;

exports[`config/loading createConfigLoader should reject if the file contents do not match the schema 1`] = `
[Error: Config from file './myconfig.ts' is malformed. Please fix the following errors:

Property 'good': Required]
`;

exports[`config/loading createConfigLoader should reject if the path is not a file 1`] = `[Error: Unable to read config file './myconfig.ts'. Check that the file exists and is readable to your terminal user]`;

exports[`config/loading createConfigLoader should reject if the path is not readable 1`] = `[Error: Unable to read config file './myconfig.ts'. Check that the file exists and is readable to your terminal user]`;

exports[`config/loading createConfigLoader should resolve if the file contents match the schema 1`] = `
{
  "good": "config",
}
`;
